Plonk: transcript hasher

vesselinux opened the pull request:
Addresses Issue #56

Commit:
plonk: replaced structure transcript_hash_t with class transcript_hasher that makes a call to hash function. for the moment it directly returns the expected hash values for the purposes of testing. addresses issue https://github.com/clearmatics/libsnark/issues/56.

dtebbs commented:
@vesselinux  please rebase onto origin/develop

vesselinux commented:
> @vesselinux please rebase onto origin/develop

Sorry, do you mean rebase onto origin/plonk?

dtebbs commented:
> > @vesselinux please rebase onto origin/develop
> 
> Sorry, do you mean rebase onto origin/plonk?

Oh yes.  origin/plonk  please.  Thanks.
I noticed that the error handling commit still appears in the history of this PR, even though that is merged to plonk now.  Rebasing should remove it (or you may have to skip it).

dtebbs commented on libsnark/zk_proof_systems/plonk/prover.tcc:
```diff
@@ -1028,27 +1017,28 @@ plonk_proof<ppT> plonk_prover<ppT>::compute_proof(
     // - beta, gamma: permutation challenges - hashes of transcript of round 1
-    const libff::Fr<ppT> beta = transcript_hash.beta;
-    const libff::Fr<ppT> gamma = transcript_hash.gamma;
+    const libff::Fr<ppT> beta = hasher.get_hash();
```
This value will be deterministic, won't it?
Should we not hash the first elements sent to the verifier?

Similarly below.  I cannot see any calls to `add_element` on the hasher.

vesselinux replied:
You are right, I am not making use of `add_element` at the moment. Will be addressed.

vesselinux commented:
> > > @vesselinux please rebase onto origin/develop
> > 
> > 
> > Sorry, do you mean rebase onto origin/plonk?
> 
> Oh yes. origin/plonk please. Thanks. I noticed that the error handling commit still appears in the history of this PR, even though that is merged to plonk now. Rebasing should remove it (or you may have to skip it).

Rebased on `plonk` and dropped the error handling commit.

dtebbs commented on libsnark/zk_proof_systems/plonk/srs.tcc:
```diff
@@ -70,19 +70,68 @@ plonk_keypair<ppT>::plonk_keypair(
-    libff::Fr<ppT> &u)
-    : beta(beta), gamma(gamma), alpha(alpha), zeta(zeta), nu(nu), u(u)
+transcript_hasher<ppT>::transcript_hasher(size_t &istep) : istep(istep)
 {
```
Would we set this to anything other than 0?  Why doesn't the constructor just create an array of the test vector values, and we read the `step`-th one?  This will also simplify `get_hash` which seems to have a lot of repeated code.

vesselinux replied:
Created an array with test values initialized by the constructor as suggested in the comment.

Regarding the `istep` variable, it has been removed since now we are using the length of the buffer as the indicator which challenge to output.

dtebbs commented on libsnark/zk_proof_systems/plonk/srs.tcc:
```diff
@@ -70,19 +70,68 @@ plonk_keypair<ppT>::plonk_keypair(
+        return u;
+    }
+    // error
+    return 0;
```
Something has gone wrong if this case, as reinforced by the comment. Let's not fail silently.  Failing silently is extremely likely to cause bugs which are very hard to track down.

vesselinux replied:
Added exception check in case of invalid buffer length plus an assert for debug.

Commit:
plonk: implementation of transcript hasher, addressing issue #56 and PR #61.

Commit:
plonk: moved the initialisation of the array of expected hash values to the transcript hasher constructor (cf. https://github.com/clearmatics/libsnark/pull/61#discussion_r935463921); added exception handling in case the buffer length is invalid (cf. https://github.com/clearmatics/libsnark/pull/61#discussion_r935473057)

Commit:
plonk: removed catching the invalid size exception within the get_hash function. the caller must handle it.

dtebbs commented on libsnark/zk_proof_systems/plonk/prover.tcc:
```diff
@@ -1060,14 +1112,20 @@ plonk_proof<ppT> plonk_prover<ppT>::compute_proof(
     // u: multipoint evaluation challenge -- hash of transcript from
     // rounds 1,2,3,4,5
-    libff::Fr<ppT> u = transcript_hash.u;
+    const libff::Fr<ppT> u = hasher.get_hash();
```
Why do we not need `u` here?  What will change when the actual hasher is implemented so that `u` is suddenly required?

vesselinux replied:
Good point. Strictly speaking, we don't need `u` to be computed here even if we are using a "real" hash function. Indeed, if you look at the paper [GWC19], pp.30, at the end of round 5 of the unrolled protocol the prover computes `u = H(transcript)` but does nothing with it. `u` is only used on the verifier side in step 9 ([GWC19], pp.31).

In my case, the reason to execute `hasher.get_hash()` at the end of the prover computation is in order to reach the maximum buffer length at which point the buffer is automatically cleared inside `get_hash()`.Upon closer look though, and prompted by your question, I now realize that there is no need to do that. The buffer can be cleared by the caller instead. The latter happens for example, when we want to re-use the same hasher object multiple times in the unit tests.

The above change is now applied and committed. In particular the computation of `u = hasher.get_hash();` is removed from the prover side.

dtebbs replied:
I'm not sure I fully follow this.  Let's discuss offline.

dtebbs replied:
Just to explain in the meantime, I think we should be hashing the entire transcript (or some derivative of it).  When the caller calls `get_hash` we could compute the hash, clear the buffer and write the new hash into the buffer, and this then allows use to call `get_hash` multiple times without adding more data (alternatively we coudl append it to the buffer).  But clearing the buffer, and allowing the calling code to do this, sounds like it could lead to problems.

W.r.t. this code, I'm fine with the prover calling `hasher.get_hash();` in order to stay in sync with verifier (as long as that's explained in a comment).  I actually think this is preferable to having having a `clear_buffer` on the transcript hasher interface (since, as discussed, it may be possible to implement a transcript_hasher without a buffer).

vesselinux replied:
I see your point, but I have a couple of clarifying questions/remarks. Let's discuss offline. It'll be faster too.

dtebbs replied:
`get_hash` may update the internal state of the transcript_hasher, depending on the implementation, therefore I think the prover and verifier must make exactly the same calls to both `add_element` *and* `get_hash`.

Let's make sure there is a comment so it's not confusing.

`libff::UNUSED(u)`

vesselinux replied:
Added back the computation of `u` in the prover. Now prover and verifier make the same number of calls to `get_hash()`.

Commit:
plonk: removed computation of the challenge u at the end of the prover process since u is not used by the prover anyway. removed the automatic clearing of the hasher buffer inside get_hash. the caller is now responsible to clear the buffer when reusing the same hasher object. see also PR comment https://github.com/clearmatics/libsnark/pull/61#discussion_r948822968 .

Commit:
plonk: use std::count to check if the hasher buffer is of valid length instead of explicitly checking all valid lengths.

dtebbs commented on libsnark/zk_proof_systems/plonk/srs.tcc:
```diff
@@ -70,17 +70,190 @@ plonk_keypair<ppT>::plonk_keypair(
+    };
+}
+
+/// clear the buffer (for now only for testing)
```
Let's keep the doc commens for functions in the header files only.

vesselinux replied:
Removed from all tcc and cpp files.

dtebbs commented on libsnark/zk_proof_systems/plonk/srs.hpp:
```diff
@@ -179,32 +179,97 @@ template<typename ppT> class plonk_keypair
+    size_t buffer_size();
+
+    // constructor
+    transcript_hasher(std::vector<uint8_t> &buffer);
```
Can you put the constructor as the first method declaration, to be consistent with other classes.

vesselinux replied:
Done

Commit:
plonk: removed doc comments /// from cpp and tcc files. left only in headers. addresses https://github.com/clearmatics/libsnark/pull/61#discussion_r949757635

Commit:
plonk: moved constructor of transcipt_hasher class as first method in the class. addresses https://github.com/clearmatics/libsnark/pull/61#discussion_r949761293

dtebbs commented:
Hasher should either keep a reference (more dangerous), or just allocate it's own internal buffer.  Latter is probably preferable.

Commit:
plonk: added back the computation of the multipoint challenge u in the last round of the prover. we need to compute u even though we are not using it in the prover in order to make sure that the prover and verifier make exactly the same number of calls to transcript_hasher.get_hash(). addresses PR #61 comment https://github.com/clearmatics/libsnark/pull/61#discussion_r952276778

Commit:
plonk: in transcript_hasher removed the private buffer variable as an input parameter to the constructor. only the object has access to the buffer now. addresses PR #61 comment https://github.com/clearmatics/libsnark/pull/61#issuecomment-1223692351

vesselinux commented:
> Hasher should either keep a reference (more dangerous), or just allocate it's own internal buffer. Latter is probably preferable.

Removed the private buffer variable as an input parameter to the constructor. Only the object has access to the buffer now.

dtebbs commented on libsnark/zk_proof_systems/plonk/prover.tcc:
```diff
@@ -9,13 +9,13 @@
 {
 
-/// Prover round 0 output constructor
+// Prover round 0 output constructor
```
This was not added as part of this PR, but I think we can remove these explanations (potentially in another PR if you prefer).  It's implicit in the function signature.

vesselinux replied:
Removed.

dtebbs commented on libsnark/zk_proof_systems/plonk/prover.tcc:
```diff
@@ -527,40 +527,40 @@ round_four_out_t<ppT>::round_four_out_t(
-/// \param[out] transcript_hasher: accumulates the communication
-///             transcript into a buffer to be hashed after prover
-///             rounds 1,2,3,4,5 (cf. fiat-shamir heuristic).
+// Prover Round 4
```
No problem having these as doc comments, but in that case, let's put them consistently in the header only.

vesselinux replied:
Moved all similar comments to hpp files only.

dtebbs commented on libsnark/zk_proof_systems/plonk/srs.hpp:
```diff
@@ -249,6 +249,9 @@ template<typename ppT> class transcript_hasher
     std::array<libff::Fr<ppT>, 6> hash_values;
 
 public:
+    // constructor
```
I think we can remove this kind of explanation.

vesselinux replied:
Removed across all files.

dtebbs commented on libsnark/zk_proof_systems/plonk/srs.tcc:
```diff
@@ -96,19 +96,19 @@ transcript_hasher<ppT>::transcript_hasher(std::vector<uint8_t> &buffer)
 }
 
-/// get buffer size
+// get buffer size
```
Again - the comment makes sense as a doc comment, so let's put it in the header instead.

dtebbs replied:
So, if someone wants to know *what* a class or method does, they look in the .hpp (at the interface and doc comments).  If they want to know *how* it does it, they look in the .tcc (at the implementation itself, and comments inside the implementation).

vesselinux replied:
Addressed across all files.

dtebbs reviewed (changes requested):
Changes w.r.t. my comments look good.  Just need to address the comments (doc comment in hpp)

Commit:
plonk: moved function signature comments from .tcc to .hpp files only; removed redundant comments indicating class constructors' other minor comments formatting

dtebbs commented on libsnark/zk_proof_systems/plonk/srs.tcc:
```diff
@@ -56,38 +54,194 @@ srs<ppT>::srs(
 }
 
-/// struct transcript_hash_t constructor
+template<typename ppT> transcript_hasher<ppT>::transcript_hasher()
```
This is basically a test vector. Ideally the main lib would define an interface and the implementation that holds these test vectors would be part of the tests dir. Further, if these test vector valves are only valid for BLS 12-381 then this test implementation wouldonly exist for that curve, not for all ppT.

dtebbs replied:
Could potentially of addressed later if you prefer to create an issue.

vesselinux replied:
Indeed these test vector values are already part of the class `plonk_example` in `tests/example.*` . I created an object of this class inside `transcript_hasher` and extracted the values from the object. 

I would like to discuss the second part i.e. making the test values available only for `BLS12-381`, before proceeding with the change.

vesselinux replied:
> if these test vector values are only valid for BLS 12-381 then this test implementation wouldonly exist for that curve, not for all ppT.

Is it true that this and also [this](https://github.com/clearmatics/libsnark/pull/61#discussion_r964178468) will be addressed after implementing the proposed steps 1-4 https://github.com/clearmatics/libsnark/pull/61#discussion_r966911185 ?

dtebbs replied:
Yes, that would address it.

dtebbs commented on libsnark/zk_proof_systems/plonk/srs.tcc:
```diff
@@ -56,38 +54,194 @@ srs<ppT>::srs(
 {
+    size_t buffer_len = this->buffer.size();
+    // DEBUG
+    printf("[%s:%d] len %7d\n", __FILE__, __LINE__, (int)buffer_len);
```
Debug code

vesselinux replied:
Removed

dtebbs commented on libsnark/zk_proof_systems/plonk/srs.tcc:
```diff
@@ -56,38 +54,194 @@ srs<ppT>::srs(
+    std::copy(str.begin(), str.end(), std::back_inserter(this->buffer));
+}
+
+template<typename ppT> libff::Fr<ppT> transcript_hasher<ppT>::get_hash()
```
Similarly, this should only be implemented far the specific curve BLS12-381

dtebbs commented on libsnark/zk_proof_systems/plonk/srs.tcc:
```diff
@@ -56,38 +54,194 @@ srs<ppT>::srs(
+        throw std::logic_error(
+            "Error: invalid length of transcript hasher buffer");
+    }
+    if (!b_valid_length) {
```
AFAICT, this should never be true because of the exception
thrown above.

vesselinux replied:
Removed

dtebbs commented on libsnark/zk_proof_systems/plonk/srs.tcc:
```diff
@@ -56,38 +54,194 @@ srs<ppT>::srs(
+            __FILE__,
+            __LINE__);
+    }
+    assert(b_valid_length);
```
similarly, I cant see how this could ever be triggered.

vesselinux replied:
Removed

dtebbs commented on libsnark/zk_proof_systems/plonk/srs.tcc:
```diff
@@ -56,38 +54,194 @@ srs<ppT>::srs(
+            __FILE__,
+            __LINE__,
+            (int)buffer_len);
+        challenge = this->hash_values[0]; // beta
```
Is the buffer really empty when the first random valve is required? This doesn't seem secure as we'll always get the same beta.

vesselinux replied:
No, it is not empty. It is of length `288` bytes -- the first value in the `length` array https://github.com/clearmatics/libsnark/blob/plonk-transcript-hasher-issue-56/libsnark/zk_proof_systems/plonk/tests/bls12_381_test_vector_transcript_hasher.cpp#L111. 

More generally, it contains the commitments to the three blinded polynomials $[a(x)]_1, [b(x)]_1, [c(x)]_1$ (Sect. 8.3, round 1 [GWC19]): https://github.com/clearmatics/libsnark/blob/plonk-transcript-hasher-issue-56/libsnark/zk_proof_systems/plonk/prover.tcc#L113-L115

dtebbs replied:
Understood.  That's my mistake - I misread the code.

dtebbs commented on libsnark/zk_proof_systems/plonk/srs.tcc:
```diff
@@ -56,38 +54,194 @@ srs<ppT>::srs(
+        challenge = this->hash_values[0]; // beta
+    }
+    // gamma
+    if (buffer_len == length[1]) {
```
The only reason to repeat all this code for each case seems to be for the variable names . If you want to keep the debug output, how about we create a vector of names.

vesselinux replied:
Addressed using a `map<size_t, string>` and a `while` loop.

dtebbs commented on libsnark/zk_proof_systems/plonk/verifier.hpp:
```diff
@@ -81,12 +80,12 @@ template<typename ppT> struct step_four_out_t {
-        libff::Fr<ppT> &&zeta,
-        libff::Fr<ppT> &&nu,
-        libff::Fr<ppT> &&u);
+        libff::Fr<ppT> &beta,
```
const

vesselinux replied:
Done

Commit:
plonk: replaced test values in transcript_hasher with values from the plonk_example class; addresses https://github.com/clearmatics/libsnark/pull/61#discussion_r964175573 .

Commit:
plonk: removed redundant debug info and unreachable debug checks. addresses PR# 61.

Commit:
plonk: set transcript hasher constants alpha, beta, ... to type const. see PR #61

dtebbs commented on libsnark/zk_proof_systems/plonk/srs.hpp:
```diff
@@ -179,32 +191,105 @@ template<typename ppT> class plonk_keypair
+///
+///     u = hash(str(buffer))
+///
+template<typename ppT> class transcript_hasher
```
This needs to be in it's own file(s).  I think we may need to:
1. create a file `/libsnark/transcript_hasher/transcript_hasher.hpp` with just a comment that describes the interface
2. add the `transcript_hasher` as a template parameter to the plonk prover and verifier. 
3. (maybe) create a `transcript_hasher_base` class which can be used as the basis for all implementations.
4. put the dummy transcript hasher (renamed to make it clear it is the implemenation for testing) in the `plonk/tests` directory.  In the tests, the prover and verifier will be instantiated as `plonk_prover<libff::bls12_381_pp, bls12_381_test_vector_transcript_hasher>` 

Later (let's ensure there is an issue for this):
6. create a `transcript_hasher` implementation using SHA3 etc, which can be used on any curve (since it does not need special constants, it can just serialize to binary and feed to SHA3).
7.  Set the default value of the `transcript_hasher` template parameter (on the plonk prover and verifier) to the sha3 hasher.

vesselinux replied:
> create a file `/libsnark/transcript_hasher/transcript_hasher.hpp` with just a comment that describes the interface

Just to confirm: file `/libsnark/transcript_hasher/transcript_hasher.hpp` will literally be empty of code and will contain only the transcript_hasher class interface as a comment -- is that your suggestion?

> put the dummy transcript hasher (renamed to make it clear it is the implementation for testing) in the `plonk/tests` directory

Do you suggest to move the current transcript hasher implementation from `srs.hpp` and `srs.tcc` to `tests/`? Would you recommend to have separate files for it or to put it directly in `tests/test_plonk.cpp`?

dtebbs replied:
> Just to confirm: file /libsnark/transcript_hasher/transcript_hasher.hpp will literally be empty of code and will contain only the transcript_hasher class interface as a comment -- is that your suggestion?

Yes.  Due to the way C++ works, we can't actually define an interface for it.  This si similar to https://github.com/clearmatics/libff/blob/develop/libff/algebra/curves/public_params.hpp#L30 and https://github.com/clearmatics/libsnark/blob/plonk-transcript-hasher-issue-56/libsnark/gadgetlib1/gadgets/pairing/pairing_params.hpp#L22, excep those files include some templated type defs.  There is a better way to do this (will discuss next time), but it's a bit involved.

The file may also contain a base class, if we create one.

> Do you suggest to move the current transcript hasher implementation from srs.hpp and srs.tcc to tests/? Would you recommend to have separate files for it or to put it directly in tests/test_plonk.cpp?

It could be in the test_plonk file, although there is probably already a lot of code in there, in which case better in it's own file.

vesselinux replied:
Addressed all steps except 3. The Issue recording the SHA3 hasher (steps 6,7) is assigned #71.

Regarding step 3:

> 3. (maybe) create a transcript_hasher_base class which can be used as the basis for all implementations.

Will this class be located in `/libsnark/transcript_hasher/transcript_hasher.hpp` ?

Commit:
plonk: added the transcript_hasher class as an additional specialization parameter to the prover and verifier classes. moved the current transcript_hasher class from files srs.* to its own files under tests, the reason being that its implementation is specific to the bls12-381 curve. set the prover and verifier to be instantiated with this bls12-381-specific implementation of transcript_hasher. this commit addresses steps 2. and 4. from comment https://github.com/clearmatics/libsnark/pull/61#discussion_r966911185 of PR#61.

Commit:
plonk: created a new transcript_hasher.hpp file with just a comment that describes the interface of a common transcript hasher. addresses step 1. from comment https://github.com/clearmatics/libsnark/pull/61#discussion_r966911185 of PR#61.

Commit:
plonk: replaced the checks over all valid buffer lengths in the transcript hasher with a while loop directly finding the correct length; addresses https://github.com/clearmatics/libsnark/pull/61#discussion_r964183723 .

dtebbs commented on libsnark/zk_proof_systems/plonk/tests/bls12_381_test_vector_transcript_hasher.cpp:
```diff
@@ -0,0 +1,157 @@
+
+    // map the index length=0,1...5 to the challenge string=beta,
+    // gamma, ...; used to print explicitly the challenge string for debug
+    std::map<size_t, std::string> challenge_str;
```
Can't this be a simple vector, initialized `{"alpha", "beta", ...}`?

dtebbs replied:
These `vector`s (`challenge_str` and `length`) can be const members of the class, and assigned in the constructor.

vesselinux replied:
> Can't this be a simple vector, initialized `{"alpha", "beta", ...}`?

Indeed. Done.

vesselinux replied:
> These `vector`s (`challenge_str` and `length`) can be const members of the class, and assigned in the constructor.

Done.

dtebbs commented on libsnark/zk_proof_systems/plonk/tests/bls12_381_test_vector_transcript_hasher.cpp:
```diff
@@ -0,0 +1,157 @@
+
+    // If we are here, then the hasher buffer has invalid length so throw an
+    // exception
+    bool b_valid_length =
```
I think this can be combined with the loop below.

vesselinux replied:
Done.

dtebbs commented on libsnark/zk_proof_systems/plonk/tests/bls12_381_test_vector_transcript_hasher.cpp:
```diff
@@ -0,0 +1,157 @@
+            "Error: invalid length of transcript hasher buffer");
+    }
+
+    libff::Fr<libff::bls12_381_pp> challenge = 0;
```
Not used until it is assigned below, at which point it is immediately returned.  Either remove (and just return `hash_values[i]`), or declare when it is assigned, so it can be `const`.

vesselinux replied:
Done.

dtebbs commented on libsnark/zk_proof_systems/plonk/tests/bls12_381_test_vector_transcript_hasher.cpp:
```diff
@@ -0,0 +1,157 @@
+        __LINE__,
+        (int)buffer_len,
+        challenge_str[i].c_str());
+    challenge = this->hash_values[i]; // beta
```
I don't think we need `this->`

vesselinux replied:
Done.

dtebbs reviewed (changes requested):
Much better structure.
Just some remaining comment about the dummy hasher code.

Commit:
plonk: replaced challenge_str map with a vector. addresses https://github.com/clearmatics/libsnark/pull/61#discussion_r968600750 .

Commit:
plonk: combine two error checks that are functionally the same into one. addresses https://github.com/clearmatics/libsnark/pull/61#discussion_r968604416 .

Commit:
plonk: declared const challenge upon assignment in  bls12_381_test_vector_transcript_hasher::get_hash(). addresses https://github.com/clearmatics/libsnark/pull/61#discussion_r968605915 .

Commit:
plonk: removed redundant this-> in get_hash. addresses https://github.com/clearmatics/libsnark/pull/61#discussion_r968607769 .

Commit:
plonk: created length and challenge arrays as const members of class bls12_381_test_vector_transcript_hasher initialized in the constructor. addresses https://github.com/clearmatics/libsnark/pull/61#discussion_r968607447 .

dtebbs commented on libsnark/zk_proof_systems/plonk/tests/bls12_381_test_vector_transcript_hasher.hpp:
```diff
@@ -0,0 +1,125 @@
+namespace libsnark
+{
+
+/// transcript hasher interface specific to the BLS12-381 curve
```
Small nit.  Something like:
```
/// Implementation of the transcript hasher interface (see transcript_hasher.hpp), returning hard-coded test vector values for the plonk proof system on a specific circuit, using the BLS12-381 curve.
```

dtebbs replied:
`Also performs a simple check on the transcript to be hashed to verify correct operation of the plonk prover and verifier.`

vesselinux replied:
Done.

dtebbs commented on libsnark/zk_proof_systems/plonk/tests/bls12_381_test_vector_transcript_hasher.hpp:
```diff
@@ -0,0 +1,125 @@
+    // hashing
+    void add_element(const libff::G2<libff::bls12_381_pp> &element);
+
+    // TODO: use following declaration to implement an actual hash function
```
Not sure what to do with this comment block, as this is test code.  Perhaps we can remove this and just keep the info in the issue?

vesselinux replied:
> Perhaps we can remove this and just keep the info in the issue?

Good suggestion. Done.

dtebbs reviewed (changes requested):
Looks good.  Just a couple of comment clarifications, and I think it's ready.

Commit:
plonk: amended comments to transcript hasher code according to suggestions in https://github.com/clearmatics/libsnark/pull/61#discussion_r970529544 .

vesselinux commented:
Question:

In `libff::Fr<libff::bls12_381_pp> bls12_381_test_vector_transcript_hasher::get_hash()` would it make sense to modify this

```
    // find the matching index
    size_t i = 0;
    while (buffer_len != length_array[i]) {
        ++i;
        if (i >= length_array.size()) {
            // If we are here, then the hasher buffer has invalid length so
            // throw an exception
            throw std::logic_error(
                "Error: invalid length of transcript hasher buffer");
        }
    }
```
as

```
    // find the matching index
    size_t i = 0;
    while (buffer_len != length_array[i])
        &&(i < length_array.size()) { ++i; }

    if (i >= length_array.size()) {
        // If we are here, then the hasher buffer has invalid length so
        // throw an exception
        throw std::logic_error(
            "Error: invalid length of transcript hasher buffer");
    }
```
?

The reason would be to prevent an infinite loop. Or would that be prevented anyway due to the exception?

dtebbs commented on libsnark/zk_proof_systems/plonk/tests/bls12_381_test_vector_transcript_hasher.cpp:
```diff
@@ -0,0 +1,138 @@
+    : length_array({288, 320, 416, 704, 896, 1120})
+    , challenge_array({"beta", "gamma", "alpha", "zeta", "nu", "u"})
+{
+    plonk_example example;
```
Let's assert that the two arrays are of the same length.

vesselinux replied:
Done.

dtebbs commented on libsnark/zk_proof_systems/plonk/tests/bls12_381_test_vector_transcript_hasher.cpp:
```diff
@@ -0,0 +1,138 @@
+    plonk_example example;
+
+    // initialize to empty vector
+    this->buffer.clear();
```
Don't need `this->` in C++.  It's required in some places to reference variables on the base class, but in general it's not needed.

vesselinux replied:
Removed. I am using `this->` to refer to own member variables of the class. Is that wrong or just redundant?

dtebbs replied:
It's just redundant and inconsistent with the rest of the code.

dtebbs commented on libsnark/zk_proof_systems/plonk/tests/bls12_381_test_vector_transcript_hasher.cpp:
```diff
@@ -0,0 +1,138 @@
+    std::copy(str.begin(), str.end(), std::back_inserter(this->buffer));
+}
+
+libff::Fr<libff::bls12_381_pp> bls12_381_test_vector_transcript_hasher::
+    get_hash()
+{
+    size_t buffer_len = this->buffer.size();
+
+    // find the matching index
+    size_t i = 0;
+    while (buffer_len != length_array[i]) {
+        ++i;
+        if (i >= length_array.size()) {
+            // If we are here, then the hasher buffer has invalid length so
+            // throw an exception
+            throw std::logic_error(
+                "Error: invalid length of transcript hasher buffer");
+        }
+    }
+
+    printf(
+        "[%s:%d] buffer_len %d: %s\n",
+        __FILE__,
+        __LINE__,
+        (int)buffer_len,
+        challenge_array[i].c_str());
+
+    const libff::Fr<libff::bls12_381_pp> challenge = hash_values[i]; // beta
+
+    return challenge;
+}
```
> The reason would be to prevent an infinite loop. Or would that be prevented anyway due to the exception?

I would just keep it simple.  Use a for loop to find the right entry, and return.  If you leave the for loop, you didn't find a match.

```suggestion
libff::Fr<libff::bls12_381_pp> bls12_381_test_vector_transcript_hasher::            
    get_hash()                                                                      
{                                                                                   
    const size_t buffer_len = this->buffer.size();                                  
                                                                                    
    // find the matching index and return the corresponding challenge.              
    for (size_t i = 0; i < length_array.size(); ++i) {                              
        if (buffer_len == length_array[i]) {                                        
            printf(                                                                 
                "[%s:%d] buffer_len %d: %s\n",                                      
                __FILE__,                                                           
                __LINE__,                                                           
                (int)buffer_len,                                                    
                challenge_array[i].c_str());                                        
                                                                                    
            return hash_values[i];                                                  
        }                                                                           
    }                                                                               
                                                                                    
    // If we are here, then the hasher buffer has invalid length so                 
    // throw an exception                                                           
    throw std::logic_error("Error: invalid length of transcript hasher buffer");    
}                                                                                   ```

vesselinux replied:
Done.

Commit:
plonk: replaced while loop in get_hash with for loop; assert-ed that length and challenge arrays are of same length in transcript hasher constructor; minor edits. addresses lates comments in PR #61 .

dtebbs commented on libsnark/zk_proof_systems/plonk/tests/bls12_381_test_vector_transcript_hasher.hpp:
```diff
@@ -0,0 +1,123 @@
+
+} // namespace libsnark
+
+#include "libsnark/zk_proof_systems/plonk/tests/bls12_381_test_vector_transcript_hasher.cpp"
```
For a .cpp we generally do not want to include it from the header, otherwise it will be compiled every time the header is included.  If the header is used in two places, we then get a link error.

It's working here because it's only used once.
Can you remove this include and add the file to the source files listed for this test in CMakeList.txt.

vesselinux replied:
Fixed. Also renamed `tests/example.tcc` to `tests/example.cpp` for consistency and added it to `CMakeList.txt` .

dtebbs commented on libsnark/zk_proof_systems/plonk/tests/bls12_381_test_vector_transcript_hasher.hpp:
```diff
@@ -0,0 +1,123 @@
+ *             (see AUTHORS).
+ * @copyright  MIT license (see LICENSE file)
+ *****************************************************************************/
+
```
On my machine (using clang), I need `#include <array>` in this file to make the build work.

vesselinux replied:
Added.

Commit:
plonk: renamed example.tcc to example.cpp. added example.cpp and bls12_381_test_vector_transcript_hasher.cpp to CMakeLists.txt. added several headers to fix compilation errors.
